src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java

```diff
@@ -0,0 +1,367 @@
+            this.physicalPk = physicalPk;
+            this.virtualHkValue = virtualHkValue;
+            this.virtualRkCondition = virtualRkCondition;
+        }
+
+        String mapKeyConditionExpression(RequestWrapper request) {
+            String physicalHkCondition = getHashKeyCondition(request);
+            String physicalRkCondition = getRangeKeyCondition(request);
+            return physicalHkCondition + " AND " + physicalRkCondition;
+        }
+
+        private String getHashKeyCondition(RequestWrapper request) {
+            AttributeValue physicalHkValue = keyMapper.toPhysicalHashKey(virtualPk.getHashKeyType(), virtualHkValue);
+            String fieldPlaceholder = setFieldPlaceholder(request, physicalPk.getHashKey());
+            String valuePlaceholder = setValuePlaceholder(request, physicalHkValue);
+            return fieldPlaceholder + " = " + valuePlaceholder;
+        }
+
+        private String getRangeKeyCondition(RequestWrapper request) {
+            // "#vhk = :vhk", no virtual RK ==> "#prk = :vhk“
+            if (virtualPk.getRangeKey().isEmpty()) {
+                return getRangeKeyEquals(request, keyMapper.toPhysicalRangeKey(virtualPk, virtualHkValue));
+            }
+
+            // "#vhk = :vhk", has virtual RK ==> "#prk BETWEEN vhk AND :vhk+FFFFFF"
+            if (virtualRkCondition.isEmpty()) {
+                AttributeValue lower = keyMapper.toPhysicalRangeKey(virtualPk, virtualHkValue, EMPTY_BYTE_ARRAY);
+                AttributeValue upper = keyMapper.toPhysicalRangeKey(virtualPk, virtualHkValue, EMPTY_BYTE_ARRAY, true);
+                return getRangeKeyBetween(request, lower, upper);
+            }
+
+            ComparisonOperator comparisonOperator = virtualRkCondition.get().operator;
+            List<AttributeValue> vrkValues = virtualRkCondition.get().values;
+
+            // "#vhk = :vhk and #vrk = :vrk" ==> "#prk = :vhk+vrk“
+            if (comparisonOperator == EQ) {
+                AttributeValue value = keyMapper.toPhysicalRangeKey(virtualPk, virtualHkValue, vrkValues.get(0));
+                return getRangeKeyEquals(request, value);
+            }
+
+            BetweenRange betweenRange = getBetweenRange(comparisonOperator, vrkValues);
+            return getRangeKeyBetween(request, betweenRange.lower, betweenRange.upper);
+        }
+
+        private BetweenRange getBetweenRange(ComparisonOperator comparisonOperator, List<AttributeValue> vrkValues) {
+            ScalarAttributeType virtualRkType = virtualPk.getRangeKeyType().get();
+            switch (comparisonOperator) {
+                case LT:
+                    // "#vhk = :vhk and #vrk < :vrk" ==> "#prk BETWEEN :vhk AND :vhk+(vrk-1)+FFFFFF“
+                    ByteBuffer rkBytes = PhysicalRangeKeyBytesConverter.toBytes(virtualRkType, vrkValues.get(0));
+                    byte[] rkByteArray = Arrays.copyOf(rkBytes.array(), rkBytes.limit());
+                    subtractOneFromByteArray(rkByteArray);
```

Comment:
This is computing the largest value that is smaller than the specified virtual hash key, since BETWEEN is inclusive, correct?

Reply:
Right that's what I'm trying to do

Reply:
So for this to be correct, there must not exist a value `v`, such that `vhk+(vrk-1)+FFFFF < v < vhk+vrk`, right? Can you explain why that's the case? (Not doubting that this is correct, just making sure I understand it.)

Reply:
Thanks for catching this! This is actually incorrect and I'm fixing it now. The algorithm should be: if the given VRK's least significant byte is 0, then strip the 0; otherwise, subtract 1 from the least significant byte.

For the case where the least significant byte is non-zero, it suffices to show that there doesn't exist a value `v` such that `(vrk-1)+FFFFFF < v < vrk` (where `vrk-1` is what we get when we subtract 1 from the last element). Suppose there is such a `v`. It's clear that all 3 sequences must share the same subsequence up to index `n-1` (1-indexed), where `n` is the length of `vrk`. At index `n`, since `v < vrk` and `vrk` has no more elements, it must be that `v[n] < vrk[n]`. So `v[n] <= vrk[n] - 1`, which means `v[n] = (vrk-1)[n]`. Then it's clear that if we fill the bytes after `(vrk-1)[n]` with the max unsigned byte, there isn't a way for `(vrk-1)+FFFFFF < v` to be true.

Reply:
Thanks, that makes sense.